### PR TITLE
[exporter/elasticsearch] handle ecs mode mapping

### PR DESCRIPTION
**Pull request by j-kap-t**
Description:

Initial pass in implementing the ecs mapping mode that is documented but not implemented. This only covers a small subset of attribute mappings from otel > ecs, but the core record fields should be properly mapped.

Link to tracking Issue:

Testing:
Added test for ecs mapping mode.

Documentation:
Updated the documentation to reflect the default behavior is the current behavior with no mapping applied.

Reopening #30454

### CODE REVIEW

File: exporter/elasticsearchexporter/model.go
Comment (gregkalapos) [inline exporter/elasticsearchexporter/model.go:57]: I think this part needs more thinking. ECS also has [`log.level`](https://www.elastic.co/guide/en/ecs/master/ecs-log.html#field-log-level) - if we deal with syslog, then this part would correctly map under `log.syslog.severity.*`, but if it's not syslog, then it should be under `log.level`.

File: exporter/elasticsearchexporter/model.go
Comment (ycombinator) [inline exporter/elasticsearchexporter/model.go:56]: Similar to the comment about `log.syslog.severity.name`, `record.SeverityNumber()` should probably map to [`event.severity`](https://www.elastic.co/guide/en/ecs/current/ecs-event.html#field-event-severity) in the general (not syslog) case?

File: exporter/elasticsearchexporter/model.go
Comment (ycombinator) [inline exporter/elasticsearchexporter/model.go:60]: `record.Body()` could be a scalar or complex value, as documented in https://github.com/open-telemetry/opentelemetry-proto/blob/9d139c87b52669a3e2825b835dd828b57a455a55/opentelemetry/proto/logs/v1/logs.proto#L167-L170.  

>   // A value containing the body of the log record. Can be for example a human-readable
  // string message (including multi-line) describing the event in a free form or it can
  // be a structured data composed of arrays and maps of other values. [Optional].
  opentelemetry.proto.common.v1.AnyValue body = 5;

So I think this conversion needs to handle those different possibilities.  Alternatively, for now, you could just check if `record.Body().Type()` is `pcommon.ValueTypeStr` and if so set `message` to `record.Body().Str()`.  We can handle the other possibilities in a follow up PR.

File: exporter/elasticsearchexporter/model.go
Comment (ycombinator) [inline exporter/elasticsearchexporter/model.go:52]: It's possible I overlooked something in ECS, but I'm not seeing a `trace.flags` field there: https://www.elastic.co/guide/en/ecs/current/ecs-tracing.html.

File: exporter/elasticsearchexporter/model.go
Comment (ycombinator) [inline exporter/elasticsearchexporter/model.go:47]: I think this logic needs to be expanded, as per https://github.com/open-telemetry/opentelemetry-proto/blob/9d139c87b52669a3e2825b835dd828b57a455a55/opentelemetry/proto/logs/v1/logs.proto#L150-L153:

```suggestion
	if record.Timestamp() != 0 {
		document.AddTimestamp("@timestamp", record.Timestamp())
	} else {
		document.AddTimestamp("@timestamp", record.ObservedTimestamp())
	}
```

**Comment by j-kap-t** [comment]
@ycombinator @gregkalapos thanks for the feedback. I hope I addressed the concerns. I wasn't the original dev on this internally, I'm just tying up some loose ends from someone who left us last year.

**Comment by ycombinator** [comment]
Thanks @j-kap-t! I reviewed the changes and they look good to me.  I'm not an approver, however, so I believe @JaredTan95 will need to review and approve this PR to get it merged.

Also, thanks for the additional context re: tying up loose ends.  Appreciate you doing that!  I'm working on a follow up PR (#31694) where we can expand upon the foundation you're laying here for `mapping.mode: ecs`.

File: .chloggen/feat_ecs-format.yaml
Comment (JaredTan95) [inline .chloggen/feat_ecs-format.yaml:4]: Others LGTM, But I have a question, is this a breaking for users who have already configured  with `ecs mode`?

Comment (j-kap-t) [reply .chloggen/feat_ecs-format.yaml:4]: @JaredTan95 thats actually a very good point. If a user explicitly has `ecs` specified in their config this would breaking change as the format of their logs would change. I will change the release notes to reflect this.

Comment (j-kap-t) [reply .chloggen/feat_ecs-format.yaml]: Fixed.

**Review by JaredTan95 — APPROVED** [review]

**Comment by j-kap-t** [comment]
@codeboten Is there anything left needed for approval & merge? Thanks!

**Comment by JaredTan95** [comment]
@j-kap-t conflicts need to be solved.

**Comment by j-kap-t** [comment]
@JaredTan95 fixed.

**Comment by ycombinator** [comment]
@open-telemetry/collector-contrib-maintainer This PR is ready to merge, whenever you get a chance.

File: .chloggen/feat_ecs-format.yaml
Comment (astencel-sumo) [inline .chloggen/feat_ecs-format.yaml:13]:
```suggestion
issues: [31553]
```

File: .chloggen/feat_ecs-format.yaml
Comment (astencel-sumo) [inline .chloggen/feat_ecs-format.yaml:10]: Why is this an "initial pass"? Why not just "Implement the `ecs` mapping mode"?

File: exporter/elasticsearchexporter/model.go
Comment (astencel-sumo) [inline exporter/elasticsearchexporter/model.go:107]: Why is this line here? Doesn't it do the same thing as the above line `m.encodeAttributes(...)`, just not taking into account the mapping mode being `raw` or `none`?

File: exporter/elasticsearchexporter/model.go
Comment (astencel-sumo) [reply exporter/elasticsearchexporter/model.go:62]: Oh I can see a previous comment about this: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31553#discussion_r1518405954.

I'm not sure if handing other body types in another PR is good enough, given that the component's stability is "Beta".

Can we at least create a separate issue to fix this after this PR is merged (hopefully soon)?

Comment (astencel-sumo) [reply exporter/elasticsearchexporter/model.go:62]: Looking more into this, it seems there might be more changes coming in future PRs to how the `ecs` mapping mode works. Is this a reasonable assumption?

If yes, I propose to add a note in the README warning users that the `ecs` mapping mode is currently not stable. This note could be removed after this mode stabilizes.

What do you think @ycombinator @j-kap-t ?

File: exporter/elasticsearchexporter/logs_exporter_test.go
Comment (astencel-sumo) [inline exporter/elasticsearchexporter/logs_exporter_test.go:188]: Looks like it's the other way around: first comes the record attributes map, second the resource attributes map.
```suggestion
			// record attrs
```

File: exporter/elasticsearchexporter/logs_exporter_test.go
Comment (astencel-sumo) [inline exporter/elasticsearchexporter/logs_exporter_test.go:193]:
```suggestion
			// resource attrs
```

File: .chloggen/feat_ecs-format.yaml
Comment (j-kap-t) [reply .chloggen/feat_ecs-format.yaml:10]: The phrasing was because this only implements a very small subset of the full ECS mapping, like this intends to expand on: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31694.

File: exporter/elasticsearchexporter/model.go
Comment (j-kap-t) [reply exporter/elasticsearchexporter/model.go:107]: Good catch, that was me failing to properly handle merge conflicts 🤦 .

File: exporter/elasticsearchexporter/model.go
Comment (ycombinator) [reply exporter/elasticsearchexporter/model.go:62]: @astencel-sumo Yes, there is at least one follow up PR in the works for `ecs` mapping mode already: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31694.  And there might be others too in the future.  So I like the idea of adding a note to the `README` warning users that `ecs` mapping mode is experimental or something to that effect as part of this PR.

**Comment by astencel-sumo** [comment]
Thanks for the updates @j-kap-t . Can you also update the README, adding a note to the `ecs` mapping mode like " :warning: This mode's behavior is unstable, it is currently undergoing changes." or something similar?

**Review by ycombinator — APPROVED** [review]

**Review by andrzej-stencel — APPROVED** [review]

**Pull request merged**
